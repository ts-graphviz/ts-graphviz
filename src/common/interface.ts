export interface IDot {
  toDot(): string;
}
