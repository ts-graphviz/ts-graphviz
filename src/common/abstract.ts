import { IDot } from '../common';
// tslint:disable: max-classes-per-file

/**
 * @category Base
 */
export abstract class GraphvizObject {}

/**
 * @category Base
 */
export abstract class DotBase extends GraphvizObject implements IDot {
  protected static quoteString(src: string): string {
    return `"${src}"`;
  }

  protected static joinLines(...lines: (string | undefined)[]): string {
    return lines.filter(l => typeof l === 'string').join('\n');
  }

  protected static indent(src: string, depth: number = 1): string {
    const indent = ' '.repeat(2 * depth);
    return src
      .split('\n')
      .map(l => `${indent}${l}`)
      .join('\n');
  }

  public abstract toDot(): string;
}
