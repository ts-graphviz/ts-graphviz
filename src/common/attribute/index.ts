export * from './attribute.js';
export * from './keys.js';
