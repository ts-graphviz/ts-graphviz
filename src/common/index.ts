export * from './type/index.js';
export * from './attribute/index.js';
export * from './models.js';
