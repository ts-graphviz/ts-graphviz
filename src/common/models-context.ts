import {
  GraphAttributesObject,
  RootGraphModel,
  SubgraphAttributesObject,
  SubgraphModel,
  NodeAttributesObject,
  NodeModel,
  EdgeTargetTuple,
  EdgeAttributesObject,
  EdgeModel,
} from './models.js';

/**
 * @beta
 */
export interface RootGraphConstructor {
  new (id?: string, attributes?: GraphAttributesObject): RootGraphModel;
  new (id?: string, strict?: boolean, attributes?: GraphAttributesObject): RootGraphModel;
  new (strict?: boolean, attributes?: GraphAttributesObject): RootGraphModel;
  new (attributes?: GraphAttributesObject): RootGraphModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): RootGraphModel;
}

/**
 * @beta
 */
export interface SubgraphConstructor {
  new (id?: string, attributes?: SubgraphAttributesObject): SubgraphModel;
  new (attributes?: SubgraphAttributesObject): SubgraphModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): SubgraphModel;
}

/**
 * @beta
 */
export interface NodeConstructor {
  new (id: string, attributes?: NodeAttributesObject): NodeModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): NodeModel;
}

/**
 * @beta
 */
export interface EdgeConstructor {
  new (targets: EdgeTargetTuple, attributes?: EdgeAttributesObject): EdgeModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): EdgeModel;
}

/**
 * @beta
 */
export interface ModelsContext {
  Graph: RootGraphConstructor;
  Digraph: RootGraphConstructor;
  Subgraph: SubgraphConstructor;
  Node: NodeConstructor;
  Edge: EdgeConstructor;
}

/**
 * @alpha
 */
export const RootModelsContext: ModelsContext = Object.seal({
  Graph: null,
  Digraph: null,
  Subgraph: null,
  Node: null,
  Edge: null,
} as unknown as ModelsContext);

/**
 * @alpha
 */
export function createModelsContext(models: Partial<ModelsContext>): ModelsContext {
  return Object.assign(Object.seal(Object.assign({}, RootModelsContext)), models);
}
