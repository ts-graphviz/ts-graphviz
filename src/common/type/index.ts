/**
 * @module ts-graphviz/type
 */
export * from './types.js';
