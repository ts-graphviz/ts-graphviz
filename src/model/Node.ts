import { DotBase } from '../common';
import { NodeAttributes } from './attributes';

/**
 * @category Primary
 */
export class Node extends DotBase {
  public readonly attributes: NodeAttributes = new NodeAttributes();
  constructor(public readonly id: string) {
    super();
  }

  public toDot(): string {
    const target = Node.quoteString(this.id);
    const attrs = this.attributes.size > 0 ? ` ${this.attributes.toDot()}` : '';
    return `${target}${attrs};`;
  }
}
