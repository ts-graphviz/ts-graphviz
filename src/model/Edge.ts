import { DotBase } from '../common';
import { EdgeAttributes } from './attributes';
import { RootClusterType } from './cluster';
import { Node } from './Node';

/**
 * @category Primary
 */
export class Edge extends DotBase {
  private get arrow(): string {
    return this.graphType === 'graph' ? '--' : '->';
  }
  public readonly attributes = new EdgeAttributes();
  private readonly nodes: Set<Node>;
  constructor(public readonly graphType: RootClusterType, node1: Node, node2: Node, ...nodes: Node[]) {
    super();
    this.nodes = new Set([node1, node2, ...nodes].filter(n => n instanceof Node));
  }

  public toDot(): string {
    const arrow = this.arrow;
    const target = Array.from(this.nodes.values())
      .map(n => Edge.quoteString(n.id))
      .join(` ${arrow} `);
    const attrs = this.attributes.size > 0 ? ` ${this.attributes.toDot()}` : '';
    const src = `${target}${attrs};`;
    return src;
  }
}
