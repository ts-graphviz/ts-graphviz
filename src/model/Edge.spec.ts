import { DotBase, GraphvizObject } from '../common';

import { Edge } from './Edge';
import { Node } from './Node';

describe('class Edge', () => {
  let edge: Edge;
  const [node1, node2, node3] = new Array(3).fill(true).map((_, i) => new Node(`node${i + 1}`));
  beforeEach(() => {
    edge = new Edge('digraph', node1, node2);
  });

  it('should be instance of Edge/DotBase/GraphvizObject', () => {
    expect(edge).toBeInstanceOf(Edge);
    expect(edge).toBeInstanceOf(DotBase);
    expect(edge).toBeInstanceOf(GraphvizObject);
  });

  describe('renders correctly by toDot method', () => {
    it('simple edge', () => {
      expect(edge.toDot()).toMatchSnapshot();
    });

    it('has some attributes', () => {
      edge.attributes.set('label', 'this is test');
      edge.attributes.set('color', 'red');
      expect(edge.toDot()).toMatchSnapshot();
    });

    it('graphType is "graph"', () => {
      edge = new Edge('graph', node1, node2);
      edge.attributes.set('label', 'this is test');
      edge.attributes.set('color', 'red');
      expect(edge.toDot()).toMatchSnapshot();
    });

    it('3 nodes', () => {
      edge = new Edge('digraph', node1, node2, node3);
      expect(edge.toDot()).toMatchSnapshot();
    });

    it('3 nodes, but many args', () => {
      edge = new Edge('digraph', node1, node2, node3, node1, node2, node3, node1, node2, node3);
      expect(edge.toDot()).toMatchSnapshot();
    });
  });
});
