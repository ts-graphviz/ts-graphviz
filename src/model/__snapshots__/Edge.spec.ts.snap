// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Edge renders correctly by toDot method 3 nodes 1`] = `"\\"node1\\" -> \\"node2\\" -> \\"node3\\";"`;

exports[`class Edge renders correctly by toDot method 3 nodes, but many args 1`] = `"\\"node1\\" -> \\"node2\\" -> \\"node3\\";"`;

exports[`class Edge renders correctly by toDot method graphType is "graph" 1`] = `
"\\"node1\\" -- \\"node2\\" [
  label = \\"this is test\\",
  color = \\"red\\",
];"
`;

exports[`class Edge renders correctly by toDot method has some attributes 1`] = `
"\\"node1\\" -> \\"node2\\" [
  label = \\"this is test\\",
  color = \\"red\\",
];"
`;

exports[`class Edge renders correctly by toDot method simple edge 1`] = `"\\"node1\\" -> \\"node2\\";"`;
